File: android/app/src/main/java/com/codergautamyt/worldguessr/WGDevShellPlugin.java
package com.codergautamyt.worldguessr;

import android.app.Activity;
import android.content.SharedPreferences;
import android.os.Handler;
import android.os.Looper;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "WGDevShell")
public class WGDevShellPlugin extends Plugin {

    @PluginMethod
    public void getUrl(PluginCall call) {
        SharedPreferences prefs = getContext().getSharedPreferences(
            MainActivity.PREFS_NAME, Activity.MODE_PRIVATE);
        String url = prefs.getString(MainActivity.PREFS_URL_KEY, null);
        JSObject ret = new JSObject();
        ret.put("url", url);
        call.resolve(ret);
    }

    @PluginMethod
    public void setUrlAndRestart(PluginCall call) {
        String url = call.getString("url");
        if (url == null || url.isEmpty()) {
            call.reject("url is required");
            return;
        }
        SharedPreferences prefs = getContext().getSharedPreferences(
            MainActivity.PREFS_NAME, Activity.MODE_PRIVATE);
        prefs.edit().putString(MainActivity.PREFS_URL_KEY, url).commit();
        call.resolve();
        restart();
    }

    @PluginMethod
    public void clearUrl(PluginCall call) {
        SharedPreferences prefs = getContext().getSharedPreferences(
            MainActivity.PREFS_NAME, Activity.MODE_PRIVATE);
        prefs.edit().remove(MainActivity.PREFS_URL_KEY).commit();
        call.resolve();
        restart();
    }

    private void restart() {
        // Tiny delay so the JS resolve gets posted before we kill the process.
        new Handler(Looper.getMainLooper()).postDelayed(() -> {
            Activity activity = getActivity();
            if (activity != null) activity.finishAffinity();
            android.os.Process.killProcess(android.os.Process.myPid());
            System.exit(0);
        }, 100);
    }
}

File: ios/App/App/WGDevShellPlugin.swift
import Foundation
import Capacitor

@objc(WGDevShellPlugin)
public class WGDevShellPlugin: CAPPlugin, CAPBridgedPlugin {

    public let identifier = "WGDevShellPlugin"
    public let jsName = "WGDevShell"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getUrl", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setUrlAndRestart", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearUrl", returnType: CAPPluginReturnPromise),
    ]

    @objc func getUrl(_ call: CAPPluginCall) {
        let url = UserDefaults.standard.string(forKey: MainViewController.prefsKey)
        call.resolve(["url": url as Any])
    }

    @objc func setUrlAndRestart(_ call: CAPPluginCall) {
        guard let url = call.getString("url"), !url.isEmpty else {
            call.reject("url is required")
            return
        }
        UserDefaults.standard.set(url, forKey: MainViewController.prefsKey)
        UserDefaults.standard.synchronize()
        call.resolve()
        restart()
    }

    @objc func clearUrl(_ call: CAPPluginCall) {
        UserDefaults.standard.removeObject(forKey: MainViewController.prefsKey)
        UserDefaults.standard.synchronize()
        call.resolve()
        restart()
    }

    private func restart() {
        // Tiny delay so the JS resolve gets posted before we kill the process.
        // exit(0) is App Store unsafe but acceptable for a dev-only IPA.
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
            exit(0)
        }
    }
}
